opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -1724,6 +1723,85 @@ namespace Opm {
                 }
                 ++local_comp_index;
             }
+
+            if (handle_ms_well && !well.segments.empty()) {
+                // we need the well_ecl_ information
+                const std::string& well_name = wm.first;
+                const Well* well_ecl = getWellEcl(well_name);
+                assert(well_ecl);
+
+                const WellSegments& segment_set = well_ecl->getWellSegments(report_step);
+
+                const int top_segment_index = state.topSegmentIndex(well_index);
+                const auto& segments = well.segments;
+
+                // \Note: eventually we need to hanlde the situations that some segments are shut
+                assert(segment_set.size() == segments.size());
+
+                for (const auto& segment : segments) {
+                    const int segment_index = segment_set.segmentNumberToIndex(segment.first);
+
+                    // recovering segment rates and pressure from the restart values
+                    state.segPress()[top_segment_index + segment_index] = segment.second.pressure;
+
+                    const auto& segment_rates = segment.second.rates;
+                    for (int p = 0; p < np; ++p) {
+                        state.segRates()[(top_segment_index + segment_index) * np + p] = segment_rates.get(phs[p]);
+                    }
+                }
+            }
+        }
+    }
+
+
+
+
+
+    template<typename TypeTag>
+    bool
+    BlackoilWellModel<TypeTag>::
+    anyMSWellOpen(const Wells* wells, const int report_step) const
+    {
+        bool any_ms_well_open = false;
```

Comment:
Would it not be easier to just loop over wells_ecl_ and call `isMultiSegment()` and `getStatus(report_step)` for each well? Or would that give something different?

Reply:
You are probably right. And also, the current approach did not consider whether it is parallel running or not, while WellState is distributed over processes. 

The difference might be related to whether it is parallel or not. Anyway, I will update it when I figure out the most proper way to do this.

Reply:
For now, looping through `wells` is still a good idea, although it does not make much difference in term of simulation. But `anyMsWellOpen` should probably be called `andMSWellOpenLocal` since it is local within the process.